expression/generator/compare_vec.go

```diff
@@ -0,0 +1,407 @@
+
+var builtinCompareVecTestCase = template.Must(template.New("").Parse(`		{retEvalType: types.ETInt, childrenTypes: []types.EvalType{types.ET{{ .ETName }}, types.ET{{ .ETName }}}},
+`))
+
+var builtinCompareVecTestFuncTail = `	},
+`
+
+var builtinCompareVecTestTail = `}
+
+func (s *testEvaluatorSuite) TestVectorizedGeneratedBuiltinCompareEvalOneVec(c *C) {
+	testVectorizedEvalOneVec(c, vecGeneratedBuiltinCompareCases)
+}
+
+func (s *testEvaluatorSuite) TestVectorizedGeneratedBuiltinCompareFunc(c *C) {
+	testVectorizedBuiltinFunc(c, vecGeneratedBuiltinCompareCases)
+}
+
+func BenchmarkVectorizedGeneratedBuiltinCompareEvalOneVec(b *testing.B) {
+	benchmarkVectorizedEvalOneVec(b, vecGeneratedBuiltinCompareCases)
+}
+
+func BenchmarkVectorizedGeneratedBuiltinCompareFunc(b *testing.B) {
+	benchmarkVectorizedBuiltinFunc(b, vecGeneratedBuiltinCompareCases)
+}
+`
+
+type CompareContext struct {
+	// Describe the name of CompareContext(LT/LE/GT/GE/EQ/NE/NullEQ)
+	CompareName string
+	// Compare Operators
+	Operator string
+}
+
+type typeContext struct {
+	// Describe the name of "github.com/pingcap/tidb/types".ET{{ .ETName }}
+	ETName string
+	// Describe the name of "github.com/pingcap/tidb/expression".VecExpr.VecEval{{ .TypeName }}
+	// If undefined, it's same as ETName.
+	TypeName string
+	// Describe the name of "github.com/pingcap/tidb/util/chunk".*Column.Append{{ .TypeNameInColumn }},
+	// Resize{{ .TypeNameInColumn }}, Reserve{{ .TypeNameInColumn }}, Get{{ .TypeNameInColumn }} and
+	// {{ .TypeNameInColumn }}s.
+	// If undefined, it's same as TypeName.
+	TypeNameInColumn string
+	// Same as "github.com/pingcap/tidb/util/chunk".getFixedLen()
+	Fixed bool
+}
+
+var typesMap = []typeContext{
+	{ETName: "Int", TypeNameInColumn: "Int64", Fixed: true},
+	{ETName: "Real", TypeNameInColumn: "Float64", Fixed: true},
+	{ETName: "Decimal", Fixed: true},
+	{ETName: "String", Fixed: false},
+	{ETName: "Datetime", TypeName: "Time", Fixed: true},
+	{ETName: "Duration", Fixed: false},
+	{ETName: "Json", TypeName: "JSON", Fixed: false},
+}
+
+var comparesMap = []CompareContext{
+	{CompareName: "LT", Operator: "<"},
+	{CompareName: "LE", Operator: "<="},
+	{CompareName: "GT", Operator: ">"},
+	{CompareName: "GE", Operator: ">="},
+	{CompareName: "EQ", Operator: "=="},
+	{CompareName: "NE", Operator: "!="},
+	{CompareName: "NullEQ"},
+}
+
+func generateDotGo(fileName string, imports string, compares []CompareContext,
+	types []typeContext) (err error) {
+	w := new(bytes.Buffer)
+	w.WriteString(header)
+	w.WriteString(newLine)
+	w.WriteString(imports)
+
+	var ctx = make(map[string]interface{})
+	for _, compareCtx := range compares {
+		for _, typeCtx := range types {
+			if typeCtx.TypeName == "" {
+				typeCtx.TypeName = typeCtx.ETName
+			}
+			if typeCtx.TypeNameInColumn == "" {
+				typeCtx.TypeNameInColumn = typeCtx.TypeName
+			}
+			ctx["compare"] = compareCtx
+			ctx["type"] = typeCtx
+			if compareCtx.CompareName == "NullEQ" {
+				// those types have done
```

Comment:
It's better to unify them into your generator.

Reply:
ok